ado/ipatrackversions.ado

```diff
@@ -0,0 +1,195 @@
+		error 101
+	}
+
+	
+	* initialize tempvars  
+	tempvar header_submit header_fvs formatted_subdate ///
+	outdated_fvs_header max_fv_by_subdate wrong_fv wrong_fv_today 
+	
+	* export sheet headers 
+	gen `header_submit' = . 
+	gen `header_fvs' = . 
+	lab var `header_submit' "Submission Date" 
+	lab var `header_fvs' "Form Versions" 	
+	export excel `header_submit' `header_fvs'  using "`saving'" in 1, sheet("T3. form versions") firstrow(varl) sheetreplace	
+
+	* convert `header_submit' to %td format if needed	
+	ds `submit', has(format %td*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = `submit'
+		}
+
+	ds `submit', has(format %tc*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofc(`submit')
+		}
+		
+	ds `submit', has(format %tC*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofC(`submit')
+		}
+		
+	ds `submit', has(format %tb*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofb(`submit')
+		}
+	
+	ds `submit', has(format %tw*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofw(`submit')
+		}
+		
+	ds `submit', has(format %tm*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofm(`submit')
+		}
+
+	ds `submit', has(format %tq*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofq(`submit')
+		}
+		
+	ds `submit', has(format %th*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofh(`submit')
+		}
+
+	ds `submit', has(format %ty*)
+	if !mi("`r(varlist)'") {
+		gen `formatted_subdate' = dofy(`submit')
+		}
+	
+	format `formatted_subdate' %tdCCYY/NN/DD
+	
+	tab `formatted_subdate' `varlist'
+	if mi("`r(N)'") {
+		di as err `"No observations in cross-tab of `submit' and `varlist' - check your data"'
+		error 122
+	}
+	
+	* format and export submission dates (left hand column of table)
+	preserve 
+	keep `formatted_subdate'
+	duplicates drop `formatted_subdate', force
+	sort `formatted_subdate' 
+	export excel using "`saving'", sheet("T3. form versions") cell(A3) sheetmodify  datestring("%tdCCYY/NN/DD") 
+	restore
+	
+	* export form def versions (column headers of table)
+	preserve 
+	table `varlist', replace
+	xpose, clear promote
+	drop in 2
+	export excel using "`saving'", sheet("T3. form versions") cell(B2) sheetmodify 
+	restore 
+
+	* export form def version counts by subdate (body of table) 
+	clear matrix
+	ta `formatted_subdate' `varlist', matcell(fvs_counts_by_subdate)
+	local num_subdates = `r(r)'
+	
+	preserve 
+	svmat fvs_counts_by_subdate, names(fvs_string)
+	keep fvs_string* 
+	export excel using "`saving'", sheet("T3. form versions") cell(B3) sheetmodify 
+	restore 
+	
+	* save max submissiondate
+	sum `formatted_subdate'
+	local max_subdate = `r(max)'
+	
+	local frmt_max_subdate: disp %tdCCYY/NN/DD `max_subdate'
+	local frmt_max_subdate = trim("`frmt_max_subdate'")
+
+	
+	* export list of obs that didn't use most recent survey version on most recent submission date 
+
+	* export header 
+	preserve
```

Comment:
So we can't do it for the table/xpose section (line 132) but did get rid of preserve/restore for the headers and the section at line 144, so I am down to two preserve/restore sections. I've been trying to play around with replacing: 

	'preserve
	keep `formatted_subdate'
	duplicates drop `formatted_subdate', force
	sort `formatted_subdate' 
	export excel using "`saving'", sheet("T3. form versions") cell(A3) sheetmodify  datestring("%tdCCYY/NN/DD") 
	restore

by trying to create a new variable, since I don't think a temp file here is actually much of an improvement. Played around with: 

    egen `tag_dates' = tag(`formatted_submit')
    gen `rdcd_formatted_submit' = `formatted_submit' if `tag_date' == 1
    export excel `rdcd_formatted_submit' using "`saving'", sheet("T3. form versions") cell(A3) sheetmodify  datestring("%tdCCYY/NN/DD")' 

but the issue is that I can't sort on this new variable without changing the main dataset....

Overall, using  two different temp files for these two sections while saving the main dataset in the background as a temp file doesn't seem like an improvement on preserve/restore but let me know if you have other ideas.